lib/pluginLoader.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+module.exports = PluginLoader;
+
+function PluginLoader (Nocca) {
+
+    var self = this;
+
+    var loadedPlugins = {};
+    var instantiatedPlugins = {};
+
+    self.registerPlugins = registerPlugins;
+    self.instantiatePlugin = instantiatePlugin;
+
+    function registerPlugins (plugins) {
+
+        plugins.forEach(function (plugin) {
+
+            if (typeof plugin === 'string') {
+                // name for plugin!
+
+                // load from the plugins dir, should return an object
+                plugin = require(__dirname + '/plugins/' + plugin);
+
+            }
+
+
+            // if plugin was an object, or was require()'d as object, proceed
+            if (typeof plugin === 'object') {
+                // plugin definition!
+
+                registerPlugin(plugin);
+
+            }
+
+        });
+
+    }
+
+    function registerPlugin (pluginDefinition) {
+
+        // make sure it's valid before continuing
+        validatePluginDefinition(pluginDefinition);
+
+        // confirm plugin is not yet loaded
+        if (typeof loadedPlugins[pluginDefinition.id] === 'undefined') {
+
+            // register plugin
+            loadedPlugins[pluginDefinition.id] = pluginDefinition;
+
+        }
+
+    }
+
+    function instantiatePlugin (pluginId) {
+
+        // bitch about it if the plugin is unknown (i.e. not defined in plugins config)
+        if (typeof loadedPlugins[pluginId] === 'undefined') {
+            throw 'Cannot use plugin with id ' + pluginId + ' as it is not loaded';
+        }
+
+        // TODO: at this time it is always a single instance that is shared. Is this desirable?
```

Comment:
Theoretically an endpoint could change the plugin used for keyGeneration. This could change (or 'use up') the instance.

Perhaps a property in the Plugin Definition Object could identify the reuse of the instance, i.e. `instance: true` or default to `singleton: true`